chisel/src/config.rs

```diff
@@ -0,0 +1,553 @@
+        &self.0
+    }
+
+    pub fn rulesets_mut(&mut self) -> &mut VecDeque<(String, Ruleset)> {
+        &mut self.0
+    }
+}
+
+impl Ruleset {
+    pub fn options(&self) -> &HashMap<String, String> {
+        &self.opts
+    }
+
+    pub fn options_mut(&mut self) -> &mut HashMap<String, String> {
+        &mut self.opts
+    }
+
+    pub fn modules(&self) -> &VecDeque<(String, ModuleConfig)> {
+        &self.modules
+    }
+
+    pub fn modules_mut(&mut self) -> &mut VecDeque<(String, ModuleConfig)> {
+        &mut self.modules
+    }
+}
+
+impl ModuleConfig {
+    pub fn options(&self) -> &HashMap<String, String> {
+        &self.0
+    }
+}
+
+impl FromArgs for ChiselConfig {
+    fn from_args(modules: &str, options: &str) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        let mut ret = VecDeque::new();
+        match Ruleset::from_args(modules, options) {
+            Ok(rs) => {
+                ret.push_back(("cli".to_string(), rs));
+                Ok(ChiselConfig(ret))
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl FromArgs for Ruleset {
+    fn from_args(modules: &str, options: &str) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        // Filter out whitespace for names.
+        let modules: String = modules.chars().filter(|c| *c != ' ').collect();
+
+        // Collect module names.
+        let module_names: Vec<String> = modules
+            .split(',')
+            .filter_map(|c| match c {
+                "" => None, // If the user put a comma somewhere weird, we need to filter out the empty string.
+                any => Some(any.to_string()),
+            })
+            .collect();
+
+        // Filter out whitespace for options.
+        let options: String = options.chars().filter(|c| *c != ' ').collect();
+
+        // Collect module options.
+        let module_options: Vec<String> = options
+            .split(',')
+            .filter_map(|c| match c {
+                "" => None,
+                any => Some(any.to_string()),
+            })
+            .collect();
+
+        let mut module_configs = VecDeque::new();
+        // Iterate over each module and generate a config from its provided options.
+        for name in module_names.iter() {
+            // Iterate over the collected options and find the options corresponding to this
+            // module.
+            let relevant_options: Vec<String> = module_options
+                .iter()
+                .filter_map(|option| {
+                    // Split by dot here to isolate the module name.
+                    let option_pair: Vec<&str> = option.split('.').collect();
+                    assert!(option_pair.len() >= 1);
+
+                    // If the module name is the same as the one we're getting options for, then
+                    // yield the option.
+                    if *name == option_pair[0] {
+                        Some(option.to_string())
```

Comment:
Since its already split here, why not return a tuple?

Reply:
I thought it would be more readable to keep the collecting relevant options part isolated from the unmarshalling step.